File: protocol.py
"""
Wire protocol for VPN packets (post-decryption).

After AES-GCM decryption, every payload starts with this 2-byte header:

    +---------+-------+----------------------------+
    | version | type  |          payload           |
    +---------+-------+----------------------------+
       1 byte  1 byte         variable

Types:
    DATA       (0)  - payload is a raw IPv4 packet to inject into TUN
    KEEPALIVE  (1)  - empty payload, used to keep NAT mappings alive
    HANDSHAKE  (2)  - initial registration: payload is client tun-IP (4 bytes)
"""

import struct

PROTOCOL_VERSION = 1

TYPE_DATA = 0
TYPE_KEEPALIVE = 1
TYPE_HANDSHAKE = 2


def pack(msg_type: int, payload: bytes = b"") -> bytes:
    return struct.pack("!BB", PROTOCOL_VERSION, msg_type) + payload


def unpack(data: bytes):
    if len(data) < 2:
        raise ValueError("Packet too short for header")
    version, msg_type = struct.unpack_from("!BB", data, 0)
    if version != PROTOCOL_VERSION:
        raise ValueError(f"Unsupported protocol version: {version}")
    return msg_type, data[2:]


def parse_ipv4_addresses(packet: bytes):
    """
    Extract (src_ip, dst_ip) from an IPv4 packet header.
    Returns a tuple of dotted-decimal strings, or (None, None) if not IPv4.
    """
    if len(packet) < 20:
        return None, None
    version = (packet[0] >> 4) & 0xF
    if version != 4:
        return None, None
    src = ".".join(str(b) for b in packet[12:16])
    dst = ".".join(str(b) for b in packet[16:20])
    return src, dst

File: android/app/src/main/java/com/vpn/Protocol.kt
package com.vpn

/**
 * Wire protocol — مطابق لـ Python vpn/protocol.py
 *
 * Post-decryption frame:
 *   [ version(1B) | type(1B) | payload ]
 *
 * Types: DATA=0, KEEPALIVE=1, HANDSHAKE=2
 */
object Protocol {
    const val VERSION: Byte = 1
    const val TYPE_DATA: Byte = 0
    const val TYPE_KEEPALIVE: Byte = 1
    const val TYPE_HANDSHAKE: Byte = 2

    fun pack(type: Byte, payload: ByteArray = ByteArray(0)): ByteArray =
        byteArrayOf(VERSION, type) + payload

    fun unpack(data: ByteArray): Pair<Byte, ByteArray> {
        require(data.size >= 2) { "Frame too short" }
        require(data[0] == VERSION) { "Unsupported protocol version: ${data[0]}" }
        return Pair(data[1], data.copyOfRange(2, data.size))
    }
}
